### DISCUSSION TITLE
Connecting to Inverter from Public Internet

### DISCUSSION THREAD

**Discussion by hulkster** [jasonacox/Powerwall-Dashboard#357] [General]
Long-time lurker who has admired this project since I started the "journey" earlier this year ... and (finally) got Solar running a month ago. FYI FWIW that I had a wrap-up interview with Project Advisor Manager and told them that _"next time you are in a meeting with Elon (LOL!), show him the Powerwall-Dashboard."_

My question is about connecting to the Inverter from the Public Internet. It appears that most people connect from "inside" their house LAN ... but I'm thinking of pulling the data from a computer "outside" my house - i.e. from the the public Internet. I run a dedicated server (running AlmaLinux9) so it would be nice to host it there.

Obviously I need to "expose" the inverter so it's accessible ... and putting it in the DMZ (i.e. full access) would not be a good idea. I'm guessing that Powerwall-Dashboard pulls everything via requests to port 80, so that's all I need to expose ... and since I'd remap it to something else (plus restrict IP address to my server), I assume that's easily changeable.

I went through the Discussion threads and didn't see anyone talk about doing this ... but can't believe I wouldn't be the first to consider it. Any comments on my idea?

Thanks!

P.S. The Tesla Inverter (7.6kW 1538000-45-A) is currently running 23.12.3 - it won't allow login via the webpage as it redirects to the _"use the Tesla Pro's App"_ screen. While using the App does work (and shows me per-string data which I'm particularly interested in), obviously something like Powerwall-Dashboard is MUCH better!

**Comment by BJReplay** [comment]
> I went through the Discussion threads and didn't see anyone talk about doing this ... but can't believe I wouldn't be the first to consider it. Any comments on my idea?

Here you go - https://github.com/jasonacox/Powerwall-Dashboard/discussions/107

**Reply by bparruck** [reply]
I think it is a bad idea, primarily being the security risk. Also, if you pulling the data every 5s, that's quite a bit of traffic. A better solution may be to host the powerwall proxy (pypowerwall container) on a small computer inside the network and create another container on the same computer to read the data from proxy and to post the data to the cloud for storage and analysis.

**Comment by jasonacox** [comment]
If you want to use a simple ssh vpn: https://github.com/jasonacox/Powerwall-Dashboard/discussions/173#discussioncomment-4889268

## Proxy the Proxy for External View

There is a way to see your Grafana dashboard from outside your home network.  It will require that you allow and NAT both port 80 for Grafana as well as port 8675 for pypowerwall.  A VPN could be set up to allow this as well and is a safer approach.  My low-budget VPN uses a locked down SSH host exposed via NAT port 7654. I use a simple script on my Mac to load the Dashboard while outside my home, like this:

```bash
#!/bin/bash

# Define the internal server IP address
SERVER=10.0.10.10
# Define the external ssh address - assumes you have already set up ssh keys
PUBLIC=user@my.home.address

# Create the tunnel
ssh -N -p 7654 $PUBLIC -L 9000:$SERVER:9000 -L 8675:$SERVER:8675

# View the Dashboard
echo "Open http://localhost:9000''

```

## Warning

Opening up Grafana or pypowerwall proxy to the public internet does come with risks.  It does mean that anyone knowing the address can see your Powerwall data.  However, more importantly, if a vulnerability exists in either of these tools, a hacker could potentially gain access to your internal network.  This is why I would recommend a VPN approach.

**Reply by jasonacox** [reply]
> "next time you are in a meeting with Elon (LOL!), show him the Powerwall-Dashboard."

🤣 Thanks @hulkster

**Comment by hulkster** [comment]
Thanks for the comments. To be clear, my situation is more similar to what @BJReplay referenced. Excellent reminder @jasonacox that one should always be concerned about security stuff ... so I would need to take steps on my server since it is already public facing.

My situation is Solar Only ... so per #183 (which I commented on with my experience to try to be helpful), it does not appear you can get String Data (from /device/vitals) ... or at least has not been figured out yet.

Speaking of being helpful, when I did the install of Docker/Grafana on CentOS7, it comes up on port3000, not 9000. Also, I think (?) Grafana changed the name from "Configured" to "Connection"  .... so if I'm not being a dumb-a** (very possible!), worth updating the excellent installation directions.

**Reply by jasonacox** [reply]
Hi @hulkster - Thanks for the note!  My test platforms are Ubuntu so I'm not sure how CentOS will work.

> when I did the install of Docker/Grafana on CentOS7, it comes up on port3000, not 9000

This is odd. Did you run the `setup.sh` script or just set them up yourself?  The 
https://github.com/jasonacox/Powerwall-Dashboard/blob/main/powerwall.yml setup pins the Grafana verion to grafana:9.1.2 and set the port to 9000. 

> Grafana changed the name from "Configured" to "Connection" 

Do you know what version of Grafana you are running?  I still see "Configuration" in the version we pin.

**Reply by hulkster** [reply]
As I said, it's very possible I'm being a "dumb-a**" ... ;-)
While I've been using Linux for decades (very familiar with chroot, etc.), it's my first-time using docker.
TL;DR is while I got things working, there's a permission issue with Grafana that I just can't figure out - GRRRR!

So here's what I did. First I remove everything and start from scratch:
_sudo yum -y remove docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin docker-ce-rootless-extras
sudo rm -rf /var/lib/containerd /var/lib/docker /etc/containerd /etc/docker /run/docker
sudo yum -y remove grafana
sudo rm -rf /var/lib/grafana /etc/grafana /run/grafana
REBOOT to clear /run_

I then install docker and grafana and enable them:
_sudo yum -y install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
sudo systemctl start docker
sudo systemctl enable docker
sudo yum -y install grafana
sudo /bin/systemctl daemon-reload
sudo /bin/systemctl start grafana-server.service
sudo /bin/systemctl enable grafana-server.service_
    NOTE: Grafana is on port 3000, NOT 9000 (which is the default when you install it)

I then run setup.sh (as a non-root user) ... and it actually "knows" when my system was installed (and Site ID) ... so that part is working fine. But NOTHING is running on port 9000 (per nmap localhost plus even "telnet localhost 9000") ... but as noted above, there is a Grafana running on port 3000 from when I installed it.

If I do a "docker logs -f grafana", I get a bunch of:
_GF_PATHS_DATA='/var/lib/grafana' is not writable.
You may have issues with file permissions, more information here: http://docs.grafana.org/installation/docker/
#migrate-to-v51-or-later_
But heck if I can figure out from that webpage what exactly I'm supposed to do to fix this ... and my Google-Kung-Fu must be getting weak in my old age as I see lots of discussion about it, but not clear-cut solution).

So when I connect to Grafana on Port3000, it says it is version 10.1.2 ... so yea, I can see in your YML file that you are specifying 9.1.2 ... so clearly I'm running what was installed on the system rather than a docker instance. However, I can still setup a dashboard ... and BOOM ... I'm seeing data!    ;-)

Note I'm Solar-ONLY ... so I'm running from the tools/solar-only directory.
BONUS points is I used tools/tesla-history and BOOM ... I've got data going back to install - SWEET!

BTW, all of the above took a dozen+ tries ... and I think (?) I got it working by adding myself to the docker and grafana group, doing a chmod 775 on /var/lib/grafana, and changing user:"1000:1000" to user:"500:498" (my UID, docker GID) ... who knows if I needed to do all that.

So EVERYTHING is working ... so while I could stay with this (?), it would be better to have a standard install. There must be something damn obvious to get grafana to work correctly using setup.sh ... since it should fire up on port 9000 and version 9.1.2.

**Reply by mcbirse** [reply]
Hey @hulkster - I am guessing since you have installed the grafana package directly on your host, this is why it is running on port 3000 (I believe it is the default?). (EDIT: Just to be clear, you should not need to install the grafana package directly)

I agree it would be better to have grafana running under docker, per the pinned version and port 9000, as was intended with the Powerwall-Dashboard project. Otherwise you are likely to come across compatibility issues or problems where you need to customise your setup to get it working correctly.

As you mentioned, it could be a permissions issue with grafana under docker.

There is a fix for this - check the file `compose.env` which allows you to set the user and group for grafana.

```bash
# Environment variables for docker compose.
#
# GRAFANAUSER - may need to be adjusted for cases such as rootless docker.
# See issue #22,
# https://github.com/jasonacox/Powerwall-Dashboard/issues/22#issuecomment-1253076441
GRAFANAUSER="1000:1000"
```

This may be the fix you are looking for?

You probably need to change the user and group to "500:498" for your system? (not sure)

Worth a try. You might have to change ownership of files under the grafana directory as well (check issue #22 for some tips).

> BONUS points is I used tools/tesla-history and BOOM ... I've got data going back to install - SWEET!

👍 🎉 - Yep, such a useful tool!! If only I got royalties on every execution, I could be rich by now. 😄

**Reply by hulkster** [reply]
OK ... I think (famous last words) I got it figured out. Grafana is now answering on port 9000 with version 9.1.2 (so the menu prompts are as documented) and I even setup OpenWeather - WOOT!    ;-)

The solution seemed to be to simply modify powerwall.yml and changes the three "user" references (two say "1000:1000" - one says  "${GRAFANAUSER}") to 500:498 ... and THEN run setup.sh.

Remember I am Solar Only, but presumably this would apply for Powerwall. 

**I REALLY appreciate the help!!!** 

FYI FWIW that this CentOS7 system is about to be retired, so I was using it for testing. I'll be installing on AlmaLinux9, so I'll report back how that goes - probably be the same.

So my step-by-step procedure is now:
_sudo yum -y remove docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin docker-ce-rootless-extras
sudo rm -rf /var/lib/containerd /var/lib/docker /etc/containerd /etc/docker /run/docker
REBOOT to clear /run  (all of the above is if you borked things up and want to start from scratch)_

_sudo yum -y install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
sudo systemctl start docker
sudo systemctl enable docker
REBOOT (although not really necessary)_

_cd Powerwall-Dashboard/tools/solar-only  (I ONLY have the inverter - no battery/powerwall)
cp powerwall.yml powerwall.yml-ORIG
Change all (three) user references in powerwall.yml to   user: "500:498"  (use your user/group - check with id)
./setup.sh (Do Open Weather)
Follow procedure to hit Web Server at port 9000 and change admin password and set things up._

_Get past history for BONUS! (have to figure out if some way to get historical weather info)
   python3 tesla-history.py --login
   python3 tesla-history.py --start "2023-08-01 00:00:00" --end "2023-09-26 10:00:00"_

**Reply by mcbirse** [reply]
Great to hear you are making progress.

> The solution seemed to be to simply modify powerwall.yml and changes the three "user" references (two say "1000:1000" - one says "${GRAFANAUSER}") to 500:498 ... and THEN run setup.sh.

I was wondering when this may become a problem. For the grafana service setup, the GRAFANAUSER variable in the compose.env file could be changed.

But, for the other services where the user/group is hard coded to 1000:1000 there is no variable defined, meaning you would need to modify the powerwall.yml file directly like you did. 

Modifying the powerwall.yml file is not recommended, as you will have issues later with upgrades, hence why a separate untracked file (compose.env) was added for the grafana user. We probably should do the same with the other service definitions. It is something I considered doing previously to avoid this problem.

> FYI FWIW that this CentOS7 system is about to be retired, so I was using it for testing. I'll be installing on AlmaLinux9, so I'll report back how that goes - probably be the same.

You may or may not have the same issue. I never had these problems (using Alpine Linux for my system) as the user/group that Powerwall-Dashboard ended up installed under was 1000:1000 anyway (with this user added to the docker group also).

> (have to figure out if some way to get historical weather info)

You can get weather history too. I wrote another tool for that as well which is in the tools directory: https://github.com/jasonacox/Powerwall-Dashboard/tree/main/tools/weather-history

Make sure to follow the instructions... you need to set up a paid subscription (however it won't cost you a cent if you limit your API calls per the instructions).

**Reply by hulkster** [reply]
Thanks yet again for the advice - some more info that may be helpful in diagnosing the root cause - LOL I did _"yum remove; yum install"_ at least another dozen times!    ;-)

First, it appears that (under tools/solar-only) that compose.env is created by setup.sh - i.e. it does not exist beforehand. Sorry I don't understand how this all works - I'm such a dinosaur that I'm an old Perl guy, rather than Python!    ;-)

For grins, I tried modifying compose.env.sample and also (as a separate test) creating compose.env and replacing GRAFANAUSER="1000:1000" with GRAFANAUSER="5555:498" (I created a new dedicated user for testing - UID is 5555 and GID is 498 (same as docker group)). All of those tests failed when running setup.sh as it complained about permission problems with tesla-history.

So after re-confirming that changing all three "user" entries (grafana, weather411, and tesla-history) in powerwall.yml works, I tried various permutations of doing less ... and it works if I only change the "user" entry for BOTH grafana and tesla-history ... does not appear I need to do for weather411. I don't know if that is helpful info (?)

Regardless, it appears to work well with those two modifications (weather data is showing) so I've got it working. So I feel bad if you are trying to troubleshoot this and it's "just me" ... although if you have something you want me to try, I'm happy to do so. Roger that I'll need to be careful to check for changes in powerwall.yml.

P.S. I just deployed on the AlmaLinux9 server. Same issue as I need to change powerwall.yml to get it to work ... plus very minor change that the GID created when installing docker is 984. One piece of good news is I'm running SELinux there ... so I was worried I would have to "wrestle" with that .. but didn't need to do anything.

PPS. Thanks for the pointer to weather-history (and writing another handy tool). I saw that, but didn't fully realize what it could do.

**Reply by mcbirse** [reply]
> So after re-confirming that changing all three "user" entries (grafana, weather411, and tesla-history) in powerwall.yml works, I tried various permutations of doing less ... and it works if I only change the "user" entry for BOTH grafana and tesla-history ... does not appear I need to do for weather411. I don't know if that is helpful info (?)

Thanks, I suspected this would be the case. The weather411 container only needs read access (can be seen in `powerwall.yml` volume section where it has `read_only: true` defined) so is probably still able to read the config file with a different or wrong user/group.

But, both grafana and the tesla-history script need write access (grafana for the db, and tesla-history for creating the .conf and .auth file and periodic updates to the .auth file). So, the user/group for those containers needs to be correct to write files in their relevant directories.

I've been meaning to get back to some updates for the Solar-Only related setup of the dashboard, as it was never quite finalised and is still listed as beta / under development... even though it is pretty much done and working for a lot of people. I've got some other pending changes for the tesla-history anyway so once I have some time to work on it will include a proposed change to address the user/group issues for the containers.

**Reply by hulkster** [reply]
I'm happy to be a Beta-tester for you @mcbirse when have something to test

**Reply by hulkster** [reply]
I did some more testing, and I think the code changes are fairly straightforward ... but disclaimer I'm a total newbie to Docker & Python, so someone knowledgeable should check. Note this is for solar-only ... may need to do something similar with pypowerwall for those with powerwalls. 

_cd tools/solar-only
cp compose.env.sample compose.env
    Change GRAFANAUSER from 1000:1000 to my UID:GID (later should be docker group)
    Duplicate that line except call it TESLAHISTORYUSER
This is what I ended up with:
   GRAFANAUSER="5555:498"
   TESLAHISTORYUSER="5555:498"_

_Modify line 66 in powerwall.yml
BEFORE: user: "1000:1000"
AFTER:    user: "${TESLAHISTORYUSER}"_


BTW, it might make sense to just declare a GENERICUSER in compose.env and use that throughout powerwall.yml, since I'm not sure much is gained by running the different services as different users (?)

setup.sh now runs without an problem. As @mcbirse pointed out, weather411 only needs read access, so there isn't a write issue with that.

**Reply by mcbirse** [reply]
I had already thought about using a single user variable, something like PWDUSER, since the Powerwall-Dashboard project likely is using the same user account for the entire project and it's docker containers.

Please feel free to raise a PR with your proposed change... or I will get to this along with other changes I have in the works once I have time.

**Reply by hulkster** [reply]
Sounds great - seems like a logical approach - maybe call it PWDUSERGROUP since it is a doublet(?)

I'll defer to people like you @mcbirse who understand the code-base better to submit the change.

Thanks for your fine work ... and bonus points if you can sneak an *option* to _"only bind to localhost"_ in the code changes per #360 ;-)
